nncf/common/quantization/quantizer_propagation/solver.py

```diff
@@ -0,0 +1,1211 @@
+            sts = self.check_branching_transition(quant_prop_graph, pq, branching_node_key)
+            if sts is TransitionStatus.SHOULD_NOT_TRANSITION:
+                abort_merge = True
+        if not abort_merge:
+            # All quantizers that are dominated by the current branching node are waiting
+            # for the merge - should merge them now
+            nncf_logger.debug("Merging PQs: {}".format(",".join([str(pq.id) for pq in waiting_pqs_list])))
+            qconfs_list = [pq.potential_quant_configs for pq in waiting_pqs_list]
+            merged_qconf_list, branch_qconf_lists = \
+                self.get_merged_qconfigs_for_downward_branching_case(qconfs_list)
+
+            if merged_qconf_list is None and \
+                    self._propagation_strategy == PropagationStrategy.MERGE_WITH_SINGLE_FQ_RESULT:
+                all_confs = '\n'.join(', '.join([f'[{str(qconf)}]' for qconf in qconfs]) for qconfs in qconfs_list)
+                nncf_logger.warning("Could not merge the quantizers at branching point {} - no common quantizer "
+                                    "configurations found among the following: \n{}".format(
+                    branching_node_key, all_confs))
+
+            merge_pqs = quant_prop_graph.merge_quantizers_for_branching_node(waiting_pqs_list,
+                                                                            merged_qconf_list,
+                                                                            branch_qconf_lists,
+                                                                            branching_node_key)
+            for idx, qconf_list in enumerate(branch_qconf_lists):
+                if qconf_list is None:
+                    merged_pqs.append(waiting_pqs_list[idx])
+                else:
+                    unmerged_pqs.append(waiting_pqs_list[idx])
+        else:
+            nncf_logger.debug("Merge aborted for PQs {}".format(",".join([str(pq.id) for pq in waiting_pqs_list])))
+            merge_pqs = []
+            unmerged_pqs = waiting_pqs_list
+
+        queue_to_cull = list(reversed(self._active_propagating_quantizers_queue))
+        self._active_propagating_quantizers_queue.clear()
+        for pq_from_queue in queue_to_cull:
+            if pq_from_queue in merged_pqs:
+                continue
+            if pq_from_queue in unmerged_pqs:
+                finished_pq = quant_prop_graph.backtrack_propagation_until_accepting_location(pq_from_queue)
+                if finished_pq is not None:
+                    self._finished_propagating_quantizers.append(pq_from_queue)
+            else:
+                self._active_propagating_quantizers_queue.appendleft(pq_from_queue)
+
+        if merge_pqs:
+            self._active_propagating_quantizers_queue.extendleft(merge_pqs)
+        self._quantizers_waiting_for_branch_merge.resolve_merged_node(branching_node_key)
+
+    def propagation_step(self, curr_prop_quantizer: PropagatingQuantizer,
```

Comment:
Could you please provide a description of the parameters? Also, please add the `:return:` section.

Reply:
DoneDone